### DISCUSSION TITLE
Will Al Assistant consider supporting Ollama or similar local storage options in the future?

### DISCUSSION THREAD

**Discussion by tyf2018** [chhoumann/quickadd#681] [General]
Will Al Assistant consider supporting Ollama or similar local storage options in the future?

Best regards.

**Comment by chhoumann** [comment]
You can already run local models😁

https://quickadd.obsidian.guide/docs/AIAssistant

**Comment by tyf2018** [comment]
> You can already run local models😁

Thank you, I understand how to use Ollama. However, there is a small issue: please tell me how to use "clipboard" this variable?When using this parameter with '<% tp.system.clipboard() %>', it will produce an error.

**Reply by chhoumann** [reply]
I don't think I understand the previous question, then. Could you clarify?

QuickAdd has its own way to handle clipboard usage: https://quickadd.obsidian.guide/docs/QuickAddAPI#getclipboard-promisestring

`<% tp.system.clipboard() %>` is a Templater thing.

What are you trying to do? I might be able to help.

**Comment by tyf2018** [comment]
😅I assumed that your plugin is compatible with the parameters of "Template".

In summary, my requirement is to have AI read content from the clipboard and respond accordingly.

Based on your usage guide, I've arranged a Clipboard2AI.js script as follows:

```
const model = "wangshenzhi/llama3-8b-chinese-chat-ollama-q8";

const settings = {
    variableName: "AnswerOfAI",
    shouldAssignVariables: false,
    modelOptions: {
        temperature: 1,
        max_tokens: 10000,
        frequency_penalty: 0.1,
        presence_penalty: 0.1
    },
    showAssistantMessages: true,
    systemPrompt: "请回答问题。"
};

const response = await api.ai.prompt(await quickAddApi.utility.getClipboard(), model, settings);
```

Unfortunately, it doesn't take effect (no reaction at all) when inserted as a macro in quickadd. Could you please provide further guidance?

![image](https://github.com/chhoumann/quickadd/assets/56342788/2c525302-3b03-4ee6-8a89-253d01765242)

**Comment by chhoumann** [comment]
Coming back to this late to close the loop on your script - sorry about that.

It does nothing because it isn't a valid QuickAdd user script yet. A user script must export a function - top-level statements are never executed, so QuickAdd finds no entry point and silently skips the file. Also, `api` and `quickAddApi` only exist as properties of the `params` object your exported function receives; they aren't globals.

Corrected version:

```js
const model = "wangshenzhi/llama3-8b-chinese-chat-ollama-q8";

module.exports = async ({ quickAddApi }) => {
  const clip = await quickAddApi.utility.getClipboard();
  await quickAddApi.ai.prompt(clip, model, {
    variableName: "AnswerOfAI",
    shouldAssignVariables: true,
    modelOptions: {
      temperature: 1,
      max_tokens: 10000,
      frequency_penalty: 0.1,
      presence_penalty: 0.1,
    },
    showAssistantMessages: true,
    systemPrompt: "请回答问题。",
  });
};
```

Two changes beyond the `module.exports` wrapper: `shouldAssignVariables: true` (you had `false`), which stores the response in the `AnswerOfAI` variable, and then a Capture or Template step after this script can insert it with `{{VALUE:AnswerOfAI}}`. Without that, the response is generated and then dropped, which also looks like "no reaction".

User script structure is covered here: https://quickadd.obsidian.guide/docs/UserScripts
